File: apps/website/src/lib/service/workspaces.ts
import type { GraphQLFetcher } from "@/lib/client";
import type {
  CreateWorkspaceInput,
  UpdateWorkspaceInput,
  Workspace,
} from "@/lib/types/workspaces";

// ---------------------------------------------------------------------------
// Queries
// ---------------------------------------------------------------------------

const WORKSPACE_FIELDS = `
  id
  userId
  name
  icon
  currency
  assetClass
  broker
  riskProfile
  totalValue
  totalValueCurrency
  createdAt
  updatedAt
  snaptradeUserId
  snaptradeConnectionId
  snaptradeAccountId
  snaptradeConnectionDisabled
  snaptradeConnectionDisabledAt
`;

const WORKSPACES_QUERY = `
  query Workspaces {
    workspaces {
      ${WORKSPACE_FIELDS}
    }
  }
`;

const WORKSPACE_QUERY = `
  query Workspace($id: String!) {
    workspace(id: $id) {
      ${WORKSPACE_FIELDS}
    }
  }
`;

// ---------------------------------------------------------------------------
// Mutations
// ---------------------------------------------------------------------------

const CREATE_WORKSPACE_MUTATION = `
  mutation CreateWorkspace($input: CreateWorkspaceInput!) {
    createWorkspace(input: $input) {
      ${WORKSPACE_FIELDS}
    }
  }
`;

const UPDATE_WORKSPACE_MUTATION = `
  mutation UpdateWorkspace($id: String!, $input: UpdateWorkspaceInput!) {
    updateWorkspace(id: $id, input: $input) {
      ${WORKSPACE_FIELDS}
    }
  }
`;

const DELETE_WORKSPACE_MUTATION = `
  mutation DeleteWorkspace($id: String!) {
    deleteWorkspace(id: $id)
  }
`;

// ---------------------------------------------------------------------------
// Service functions
// ---------------------------------------------------------------------------

export async function fetchWorkspaces(
  fetcher: GraphQLFetcher,
): Promise<Workspace[]> {
  const data = await fetcher<{ workspaces: Workspace[] }>(WORKSPACES_QUERY);
  return data.workspaces;
}

export async function fetchWorkspace(
  fetcher: GraphQLFetcher,
  id: string,
): Promise<Workspace | null> {
  const data = await fetcher<{ workspace: Workspace | null }>(WORKSPACE_QUERY, {
    id,
  });
  return data.workspace;
}

export async function createWorkspace(
  fetcher: GraphQLFetcher,
  input: CreateWorkspaceInput,
): Promise<Workspace> {
  const data = await fetcher<{ createWorkspace: Workspace }>(
    CREATE_WORKSPACE_MUTATION,
    { input },
  );
  return data.createWorkspace;
}

export async function updateWorkspace(
  fetcher: GraphQLFetcher,
  id: string,
  input: UpdateWorkspaceInput,
): Promise<Workspace> {
  const data = await fetcher<{ updateWorkspace: Workspace }>(
    UPDATE_WORKSPACE_MUTATION,
    { id, input },
  );
  return data.updateWorkspace;
}

export async function deleteWorkspace(
  fetcher: GraphQLFetcher,
  id: string,
): Promise<boolean> {
  const data = await fetcher<{ deleteWorkspace: boolean }>(
    DELETE_WORKSPACE_MUTATION,
    { id },
  );
  return data.deleteWorkspace;
}

File: backend/src/service/read_service/workspaces.rs
use anyhow::Result;

use crate::service::db::client::UserDb;
use crate::service::db::schema::tables::workspaces_table::{
    self, CreateWorkspaceInput, UpdateWorkspaceInput, Workspace,
};

pub async fn list_workspaces(user_db: &UserDb) -> Result<Vec<Workspace>> {
    workspaces_table::list_workspaces(user_db.pool(), user_db.user_id()).await
}

pub async fn get_workspace(user_db: &UserDb, id: &str) -> Result<Option<Workspace>> {
    workspaces_table::find_workspace(user_db.pool(), id, user_db.user_id()).await
}

pub async fn create_workspace(user_db: &UserDb, input: CreateWorkspaceInput) -> Result<Workspace> {
    workspaces_table::create_workspace(user_db.pool(), user_db.user_id(), input).await
}

pub async fn update_workspace(
    user_db: &UserDb,
    id: &str,
    input: UpdateWorkspaceInput,
) -> Result<Workspace> {
    workspaces_table::update_workspace(user_db.pool(), id, user_db.user_id(), input).await
}

pub async fn delete_workspace(user_db: &UserDb, id: &str) -> Result<bool> {
    workspaces_table::delete_workspace(user_db.pool(), id, user_db.user_id()).await
}
